Added FSDP checkpoint handling to convert_torch_dist_to_hf.py

cklxx opened the pull request:
Summary

This PR adds a dedicated converter for FSDP backend checkpoints (saved via torch.distributed.checkpoint, typically without common.pt) and documents how to use it to export a Hugging Face–compatible model.

Motivation

fix https://github.com/THUDM/slime/issues/1094

Below is the script I tested locally.
```
FSDP model loaded in 7.26 sec.
Missing keys: []
Unexpected keys: []
Model weights saved to /tmp/slime_fsdp2hf_smoke_20251214_140145/out_hf
copy from /tmp/slime_fsdp2hf_smoke_20251214_140145/origin_hf/config.json to /tmp/slime_fsdp2hf_smoke_20251214_140145/out_hf/config.json
copy from /tmp/slime_fsdp2hf_smoke_20251214_140145/origin_hf/chat_template.jinja to /tmp/slime_fsdp2hf_smoke_20251214_140145/out_hf/chat_template.jinja
copy from /tmp/slime_fsdp2hf_smoke_20251214_140145/origin_hf/tokenizer_config.json to /tmp/slime_fsdp2hf_smoke_20251214_140145/out_hf/tokenizer_config.json
copy from /tmp/slime_fsdp2hf_smoke_20251214_140145/origin_hf/special_tokens_map.json to /tmp/slime_fsdp2hf_smoke_20251214_140145/out_hf/special_tokens_map.json
copy from /tmp/slime_fsdp2hf_smoke_20251214_140145/origin_hf/tokenizer.json to /tmp/slime_fsdp2hf_smoke_20251214_140145/out_hf/tokenizer.json
/usr/local/lib/python3.12/dist-packages/transformers/utils/hub.py:110: FutureWarning: Using `TRANSFORMERS_CACHE` is deprecated and will be removed in v5 of Transformers. Use `HF_HOME` instead.
  warnings.warn(
✅ Converted model load+forward OK
logits shape: (1, 1, 59264)
OUT_DIR: /tmp/slime_fsdp2hf_smoke_20251214_140145/out_hf
✅ ALL DONE. Artifacts kept in: /tmp/slime_fsdp2hf_smoke_20251214_140145
root@e77bc5c24070:/workspace# ls /tmp/slime_fsdp2hf_smoke_20251214_140145
fsdp_ckpt  hf_home  origin_hf  out_hf
```

PopSoda2002 reviewed (commented):
Hi, it seems your PR description and implementation have mismatch, can you double check that?

cklxx commented:
@PopSoda2002 
Sorry about that — I thought adding a dedicated script would be more appropriate, so I changed the implementation accordingly. I’ve updated the PR description as well. Thanks a lot for the review!

cklxx commented on tools/convert_fsdp_to_hf.py:
```diff
@@ -0,0 +1,134 @@
+        is_coordinator: bool = False,
+    ) -> None:
+        for k, v in metadata.state_dict_metadata.items():
+            if "optimizer" in k or "_state" in k:
```
Thank you very much for the correction. It has been fixed — hope it helps!

cklxx commented on tools/convert_fsdp_to_hf.py:
```diff
@@ -0,0 +1,134 @@
+    print(f"FSDP model loaded in {time.time()-t:.2f} sec.")
+
+    model_state = {
+        k.replace("model_state.model.", "", 1).replace("model.", "", 1): v
```
All set!

cklxx commented:
@PopSoda2002 Please help review this PR. I will quickly make improvements and move it forward. Thank you very much!

PopSoda2002 commented on tests/test_convert_fsdp_to_hf.py:
Hi, I think we can delete this test

PopSoda2002 reviewed (approved):
Thanks for the contributing! LGTM, but can @Hecate0821 also reviews this PR to double check/

cklxx commented:
@Hecate0821 Could you please help take a look at this PR when you have time? Thanks a lot!

cklxx commented:
@PopSoda2002 @Hecate0821  Thanks for the review! To address your questions: I think --force is necessary, as it’s common to convert a trained model and overwrite the existing one. As for --prefix, it was introduced due to issues encountered during earlier changes, but I agree it’s no longer needed and have removed it. And it has been verified locally.

cklxx commented:
@PopSoda2002 @Hecate0821 Sorry for the bother — i noticed that there are 3 workflows awaiting approval. Is there anything I should do on my side to proceed??
